### ISSUE TITLE
Having issues with receiving data on the other end with this library specifically

### ISSUE THREAD

**Issue by CountryTk** [jhelovuo/RustDDS#236]
I have a .idl file that defines the structs i can send to the receiving end and when doing it with this library, the message doesn't make it.
I used the same idl and set it up using FastDDS and the message went through.

QoS was the same, topic name, desc was the same and the payload was the same.

Could you point me in the right direction on where to start debugging it? I looked on wireshark and it's getting sent, just not being received 

I could give more context over email

**Comment by jhelovuo** [comment]
Please describe in more detail which DDS libraries are you using to send and receive, and which combinations work, and which do not.

Are you changing between RustDDS and FastDDS senders and receiver is fixed? (What is the receiver?)

In addition to Wireshark, you can set up a logger to see what RustDDS is doing. Please see an example in [the shapes demo](https://github.com/jhelovuo/RustDDS/blob/master/examples/shapes_demo/main.rs#L318). Configure the rustdds::dds::reader (or ::writer) component log level to "info" or "debug". You should then see from the logs if RustDDS thinks it is matching with the reader/writer at the opposite end, or is there a QoS mismatch. Note that this does not still guarantee that the other end thinks the same.

**Comment by CountryTk** [comment]
Receiver is IntercomDDS. 

When sending a heartbeat message to the receiver via RustDDS then it works, it receives the message. After a heartbeat is established between the 2 systems (RustDDS and IntercomDDS) it is then possible to send another message, let's call it Start_Order. After sending a valid Start_Order to the receiver, the receiver is not getting the message. 

When I used FastDDS the receiver stayed the same and when I created the Start_Order message there, then it went through and made it to IntercomDDS.

I'll setup the logger for RustDDS and see if it reports any missing QoS or something.

But also as an added note: the heartbeat message has no enums in it and it goes through, the start_order message has a few enums and it doesn't go through, maybe somewhere an enum is getting serialized in the wrong way and the data is wrong so IntercomDDS just outright rejects it? Not sure, I'm very new to DDS and RTPS myself.

Thank you for help!

**Comment by CountryTk** [comment]
Okay, so the logger is telling me that after I send the Start_order message

Topic "Start_Order" update ignored from GUID {[...]} Already have this.

But nothing about wrong QoS or anything about why the message is not making it to IntercomDDS.

**Comment by CountryTk** [comment]
Since the data has the CDR_LE output encoding, maybe it's somehow encodes the enums wrong? For example in the .idl file I have this

enum Type_E {
TYPE_NO_STATEMENT,
TYPE_SOMETHING_ELSE,
TYPE_TIME
}

and in rust I have an enum for that like this:

#[derive(Serialize, Deserialize, Clone, Debug, Copy)]
pub enum Type_E {
    TYPE_NO_STATEMENT,
    TYPE_SOMETHING_ELSE,
    TYPE_TIME
}

**Comment by CountryTk** [comment]
When sending the Start_Order message to IntercomDDS with FastDDS and then starting RustDDS then the logger tells me this:

ERROR rustdds::dds::with_key::datareader Failed to deserialize bytes: unexpected end of input, Topic = Start_Order, Type = TypeDesc {my_name: "Start_Order"}

**Comment by jhelovuo** [comment]
Ah, ok. This seems to be some disgreement over enum or tagged union serialization. Enums are not a much tested feature in RustDDS, so there may be a bug lurking there. Please note that `enum` in IDL / CDR means about the same as it does in the C language. The meaning of `enum` in Rust is much more general.

Can you provide some minimal example that does not work so I could reproduce it?

**Comment by CountryTk** [comment]
Okay so it seems like the enums are being serialized correctly. I setup the shapes demo with fastdds as the subscriber and rustdds as the publisher. Added a enum to the Shapes struct on both ends and the message was received successfully and deserialized as 32-bit integer.

The deserialization error comes after these steps are taken
1) Send a message with FastDDS to IntercomDDS
2) Close FastDDS and open RustDDS and start listening on events
3) IntercomDDS sends data to RustDDS and that gives the deserialization error.

The endianness bit is set. When deserializing a response that is sent by IntercomDDS upon receiving the message I send, I SOMETIMES get a "attempt to subtract with overflow" panic.

It's really hard to provide a minimal working example because the IntercomDDS part is proprietary and when I couldn't reproduce it with a modified Shapes example.

**Comment by jhelovuo** [comment]
> When deserializing a response that is sent by IntercomDDS upon receiving the message I send, I SOMETIMES get a "attempt to subtract with overflow" panic.

This is really something new. Can you provide a backtrace of this?

> It's really hard to provide a minimal working example because the IntercomDDS part is proprietary and I couldn't reproduce it with a modified Shapes example.

This will be quite difficult to advice on then. Can you tell which kind of data types are used? Strings? Byte sequences? Fixed-length arrays? Unions? Structs within structs within unions?

Maybe you could give the definition of the offending data type but just replace all field names with "field1, "field2" and type names with "T1", T2", etc. so you do not reveal details of the application ?

**Comment by CountryTk** [comment]
I did some more research and made FastDDS act as the subscriber of the data I want to send with RustDDS and it didn't receive it.
I used fastddsgen to generate files from my idl file, could you recommend me a tool that does that for rust? The one I found online failed with the idl because the idl contains @key attributes in some structs etc.

I also recall when trying to get the shape demo working that I had to add @key string shape to the .idl I created with fastddsgen to get it working with RustDDS.

So basically I think the issue (hopefully) lies in the structs I manually built myself following the given .idl file I have.

I'm going to try and build a minimal working example with FastDDS as the subscriber and RustDDs as the publisher tomorrow but before I build that I wanna make sure it's not my structs that are wrong

**Comment by jhelovuo** [comment]
Currently, you have to do mapping of IDL to Rust structs by hand. It is quite straightforward. Ints map to ints, floats to floats, strings to strings, etc. Fixed-size IDL array maps to fixed-size Rust array. Variable-length sequence in IDL maps to Vec.

The key for DDS topics with key in RustDDS is not defined as an attribute of struct fields, but it is defined by implementing the trait Keyed on your message (data sample) type. They Keyed trait defines what key type is associated with that data sample type and how to extract the key from a data sample.

**Comment by CountryTk** [comment]
Okay, the keyed trait might be it. In my .idl file for the struct i mapped to rust there are 2 keys
struct my_order
{
    @key int resourceId;
    @key int anotherId;
}

I tried implementing keyed trait for each of them and it didn't work, how would I go about implementing it for BOTH of them?

**Comment by jhelovuo** [comment]
There is no double key, but a single key that consists of two fields. 

You would go about implementing two-field key (more or less) as follows: 

```
impl Keyed for MyOrder {
  type K = (u32,u32);
  fn key(&self) -> (u32,u32) {
    (self.resourceid , self.anotherId)
  }
}
```

Please try something like the above. Naturally, use the correct integer type if `int` maps to something else than `u32`. 

Alternatively, you could define a type
```
struct MyOrderKey {
  resourceId: u32,
  anotherId: u32,
}
```

and then use MyOrderKey in place of `(u32,u32)`.

Keys should be sent over RTPS only when disposing data samples in a with_key topic. You could catch this with Wireshark. Look for DATA submessages where flags are  DataFlag=false, KeyFlag=true. The payload should then be the serialized key. 

Some RTPS implementations use an alternative method, and set DataFlag=false, KeyFlag=false, InlineQoSFlag=true. Then the inline QoS part of the submessage should contain a parameter KEY_HASH with a hash of the key, or in the case of short keys, just the key itself.

**Comment by CountryTk** [comment]
The first solution tells me that Trait Key is not implemented for (u32, u32) and the second one gives the same error but for MyOrderKey

I tried implementing Key trait on the MyOrderKeyStruct but I'm not sure where I can access CdrEncodingSize derive macro

**Comment by CountryTk** [comment]
I added the cdr encoding size dependency to my cargo toml file and I seem to have gotten it working, thank you for the help!
